### PR TITLE
feat: result pragma support

### PR DESCRIPTION
**Pull request by aniksd-braket**
# Add support for Braket result type pragmas
  
  ## Summary
  
  Adds end-to-end support for Braket result type pragmas (`#pragma braket result ...`) in the `to_qiskit()` → `to_braket()` pipeline. Previously, any OpenQASM program containing result pragmas would crash
  with `NotImplementedError`. Now they are parsed, stored as circuit metadata, and processed into the correct basis rotation gates and measurements.
  
  ## Changes
  
  **`qasm_context.py`** — Parse and store result pragmas
  - Override `parse_pragma()` to capture the raw pragma text
  - Override `add_result()` to store parsed IR objects as circuit metadata
  - Attach `braket_result_pragmas` to circuit metadata on access
  
  **`passes/basis_rotation_pass.py`** — New `AddBasisRotationGates` TransformationPass
  - Reads result pragma metadata from the circuit
  - Appends appropriate basis rotation gates per observable:
    - Z, I → no rotation
    - X → H
    - Y → Z, S, H
    - H → Ry(-π/4)
    - Hermitian → Unitary (eigenvector matrix)
  - Appends measurements on all targeted qubits
  - Skips basis-invariant types (state_vector, density_matrix, amplitude)
  
  **`compilation.py`** — Integrate into `_compile()` pipeline
  - Detects circuits with result pragma metadata after transpilation
  - Runs `AddBasisRotationGates` pass automatically
  
  ## Testing
  
  - **16 tests** — Parsing and metadata attachment in `to_qiskit()`
  - **14 tests** — `AddBasisRotationGates` pass unit tests (all observable types, edge cases)
  - **15 tests** — End-to-end through `to_braket()` with full OQ3 string assertions and metadata verification
  
  ## Result type support matrix
  
  | Result Type | Behavior |
  |------------|----------|
  | Expectation, Sample, Variance | Basis rotation gates + measurement |
  | Probability | Measurement only (Z-basis) |
  | StateVector, DensityMatrix, Amplitude | No gates or measurements added |

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ ] I have read the [CONTRIBUTING](https://github.com/amazon-braket/qiskit-braket-provider/blob/main/CONTRIBUTING.md) doc
- [ ] I used the PR title format described in [CONTRIBUTING](https://github.com/amazon-braket/qiskit-braket-provider/blob/main/CONTRIBUTING.md#pr-title-format)
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/qiskit-braket-provider/blob/main/README.md) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: qiskit_braket_provider/providers/compilation.py [hunk 1]
```diff
@@ -231,6 +237,13 @@ def _compile(
     if isinstance(target, _SubstitutedTarget):
         circuits = target._substitute(circuits)
 
+    has_result_pragmas = any(
+        (circ.metadata or {}).get("braket_result_pragmas") for circ in circuits
+    )
+    if has_result_pragmas and not preserve_result_pragmas:
+        basis_rotation_pm = PassManager([AddBasisRotationGates()])
```
Comment (nilsquet) [inline qiskit_braket_provider/providers/compilation.py:244]: This compilation pass must run before the actual compilation. Right now, 1) the added rotation gates are not nativized and 2) the applied mapping is not considered.

File: qiskit_braket_provider/providers/passes/basis_rotation_pass.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            return dag
+
+        num_qubits = dag.num_qubits()
+        qubits_to_measure = set()
+        rotation_ops = []
+
+        for pragma_entry in result_pragmas:
+            parsed = pragma_entry["parsed"]
+
+            if isinstance(parsed, _BASIS_INVARIANT_TYPES):
```
Comment (nilsquet) [inline qiskit_braket_provider/providers/passes/basis_rotation_pass.py:97]: `state_vector`, `density_matrix`, and `amplitude` don't need rotation or measurement, so skipping them here is right. But the pragma only lives on `circuit.metadata["braket_result_pragmas"]` at this point, and nothing downstream (qasm3.dumps, to_braket) reads that metadata back. The result is OQ3 output with no pragma and no measurement — the simulator has no result type to compute.

File: qiskit_braket_provider/providers/passes/basis_rotation_pass.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    try:
+        np_matrix = np.array([[complex(c[0], c[1]) for c in row] for row in matrix])
+    except (IndexError, TypeError, ValueError) as e:
+        raise ValueError(f"Invalid Hermitian matrix format in result type pragma: {e}") from e
+    _eigenvalues, eigenvectors = np.linalg.eigh(np_matrix)
+    unitary = eigenvectors.conj().T
+    return [(UnitaryGate(unitary), targets)]
+
+
+class AddBasisRotationGates(TransformationPass):
```
Comment (nilsquet) [inline qiskit_braket_provider/providers/passes/basis_rotation_pass.py:64]: The naming is misleading since this function also adds measurements.

File: qiskit_braket_provider/providers/passes/basis_rotation_pass.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+                            qubits_to_measure.update(obs_targets)
+                            obs_idx += num_qubits_for_obs
+
+        if not qubits_to_measure and not rotation_ops:
+            return dag
+
+        num_clbits_needed = len(qubits_to_measure)
+        clbits_to_add = num_clbits_needed - dag.num_clbits()
+        if clbits_to_add > 0:
+            for _ in range(clbits_to_add):
```
Comment (nilsquet) [inline qiskit_braket_provider/providers/passes/basis_rotation_pass.py:141]: If the source already had bit[N] c; or explicit measure, the pass reuses those clbits and adds new measurements on top, producing silent double-measurement.

File: qiskit_braket_provider/providers/passes/basis_rotation_pass.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                dag.add_clbits([Clbit()])
+
+        for gate, target in rotation_ops:
+            if isinstance(target, list):
+                dag.apply_operation_back(gate, [dag.qubits[t] for t in target])
+            else:
+                dag.apply_operation_back(gate, [dag.qubits[target]])
```
Comment (nilsquet) [inline qiskit_braket_provider/providers/passes/basis_rotation_pass.py:148]: target and qubit here originate from parsed.targets (source-level indices like q[0]→0). After transpile, dag.qubits[i] is the qubit at DAG slot i, not the source's q[i]. Rotations and measurements land on the wrong qubit whenever the layout permutes.

Should be fine if this pass is executed before the compilation.

File: qiskit_braket_provider/providers/passes/basis_rotation_pass.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    Variance,
+)
+
+_BASIS_INVARIANT_TYPES = (StateVector, DensityMatrix, Amplitude)
+_Z_BASIS_TYPES = (Probability,)
+_OBSERVABLE_TYPES = (Expectation, Sample, Variance)
+
+
+def _rotation_gates_for_observable(observable: str, target: int) -> list[tuple]:
+    """Return (gate, qubit) pairs for rotating from the given observable basis to Z basis."""
```
Comment (nilsquet) [inline qiskit_braket_provider/providers/passes/basis_rotation_pass.py:27]: Inconsistent docstring style compared to `_rotation_gates_for_hermitian` where args and returns is explicitly stated.

File: qiskit_braket_provider/providers/qasm_context.py [hunk 7]
```diff
@@ -132,6 +135,41 @@ def circuit(self) -> QuantumCircuit:
             )
         return self._circuit_stack[0]
 
+    def parse_pragma(self, pragma_body: str):  # noqa: ANN202
```
Comment (nilsquet) [inline qiskit_braket_provider/providers/qasm_context.py:138]: Why is `# noqa: ANN202` needed? Adding a return type should fix it. Also, the docstring should document what's returned.

File: qiskit_braket_provider/providers/qasm_context.py [hunk 8]
```diff
@@ -132,6 +135,41 @@ def circuit(self) -> QuantumCircuit:
...
+
+        Overrides AbstractProgramContext.parse_pragma() to store the raw pragma
+        command string before parsing. This allows us to reconstruct the full
+        pragma text for re-emission in the compiled output.
+
+        Args:
+            pragma_body: The body of the pragma statement (e.g., "braket result expectation z(q[0])").
+        """
+        self._last_pragma_command = pragma_body
+        return super().parse_pragma(pragma_body)
```
Comment (nilsquet) [inline qiskit_braket_provider/providers/qasm_context.py:149]: `parse_pragma` override and `_last_pragma_command` field only exist to smuggle raw text over to `add_result` so it can build `raw_pragma`. But `raw_pragma` isn't consumed anywhere in this PR — it's written to metadata and never read. Drop both.

If a future consumer needs the pragma text, reconstruct it from `result` at that point (`braket.circuits.result_types.*` classes already implement `_to_openqasm`).

Bonus: the `"adjoint_gradient" in self._last_pragma_command` substring check becomes `isinstance(result, AdjointGradient)` — precise, and no longer needs the raw text.

File: qiskit_braket_provider/providers/passes/basis_rotation_pass.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    Expectation,
+    Probability,
+    Sample,
+    StateVector,
+    Variance,
+)
+
+_BASIS_INVARIANT_TYPES = (StateVector, DensityMatrix, Amplitude)
+_Z_BASIS_TYPES = (Probability,)
+_OBSERVABLE_TYPES = (Expectation, Sample, Variance)
```
Comment (nilsquet) [inline qiskit_braket_provider/providers/passes/basis_rotation_pass.py:23]: Should this be moved to `gate_mappings.py` where we store all other constants? Admittedly, the file name is not suitable and should rather be called something like "constants" or "common".

File: qiskit_braket_provider/providers/passes/basis_rotation_pass.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    measurements are added. For observable types (expectation, sample, variance),
+    rotation gates are added before measurements.
+
+    AdjointGradient is not supported by this pass — it requires the pragma to be
+    preserved verbatim for simulator-side processing. Programs containing
+    adjoint_gradient pragmas are not parseable by the default simulator's pragma
+    grammar and will not reach this pass.
+    """
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
```
Comment (nilsquet) [inline qiskit_braket_provider/providers/passes/basis_rotation_pass.py:82]: The pass does a lot in one method: type dispatch (basis-invariant / Z-basis / observable), target resolution (None → all), observable-form dispatch (single string vs. list vs. hermitian matrix), tensor-product bookkeeping via obs_idx, deferred measurement collection, clbit allocation, and DAG mutation.
  Each branch is small on its own, but the interactions are where the subtle behavior lives.

Suggest splitting: a small pure helper per result-type that returns a (rotation_gates, qubits_to_measure) plan from a parsed object, and a thin pass body that unions plans and applies them to the DAG. Easier to reason about and to test.

**Review by nilsquet — CHANGES_REQUESTED** [review]
Thanks for the thorough work here — two high-level notes on the test suite:

1. Feels heavy: ~900 lines of tests for ~200 lines of source, with each result type exercised at the parse, pass, and end-to-end layers. Suggest collapsing to end-to-end per result type plus a small unit-test file for edge cases. Given the volume, I didn't review the test bodies line-by-line.
2. to_braket is always called without `target`, `braket_device`, or `basis_gates`. Therefore, the test don't show that the added rotation gates are not compiled.

File: qiskit_braket_provider/providers/qasm_context.py (hunk 8 above)
Comment (laurencap) [reply qiskit_braket_provider/providers/qasm_context.py:149]: +1 to this. It's always possible to add back later if it ends up being needed elsewhere.

File: qiskit_braket_provider/providers/passes/basis_rotation_pass.py (hunk 2 above)
Comment (aniksd-braket) [reply qiskit_braket_provider/providers/passes/basis_rotation_pass.py:97]: Yup, that's a valid concern. I will make sure to make sure that the metadata is appropriately passed to downstream simulators and devices.

File: qiskit_braket_provider/providers/passes/basis_rotation_pass.py (hunk 4 above)
Comment (aniksd-braket) [reply qiskit_braket_provider/providers/passes/basis_rotation_pass.py:141]: added to guard to make sure a circuit with result pragmas does not have pre-existing measurements

**Comment by laurencap** [comment]
This has been broken into smaller PRs which have been merged! 🎉  Closing.
